From ad5e69629ebcec7615745142af5f6456ef942e80 Mon Sep 17 00:00:00 2001
From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Fri, 7 Aug 2026 07:53:29 +0200
Subject: Revert "x86/bugs: Make Safe-RET robust against interrupt injection"

This reverts commit 9de1a49e8f1fbf7c372902573a27a97d4ab4d0af which is
commit 7e7f81cf6f5ca3311e526308f55d7c54d3ba71f9 upstream.

It was incorrect, a more correct fix will be applied next.

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 arch/x86/entry/entry_64.S            |  5 ----
 arch/x86/include/asm/nospec-branch.h | 56 ------------------------------------
 arch/x86/kernel/cpu/bugs.c           | 39 -------------------------
 arch/x86/lib/retpoline.S             | 20 -------------
 4 files changed, 120 deletions(-)

diff --git a/arch/x86/entry/entry_64.S b/arch/x86/entry/entry_64.S
index 098dfdfe99f73..bd785386d6291 100644
--- a/arch/x86/entry/entry_64.S
+++ b/arch/x86/entry/entry_64.S
@@ -971,8 +971,6 @@ SYM_CODE_START_LOCAL(paranoid_entry)
 	IBRS_ENTER save_reg=%r15
 	UNTRAIN_RET
 
-	HANDLE_INTR_SAFERET 8(%rsp)
-
 	RET
 SYM_CODE_END(paranoid_entry)
 
@@ -1080,9 +1078,6 @@ SYM_CODE_START_LOCAL(error_entry)
 	movl	%ecx, %eax			/* zero extend */
 	cmpq	%rax, RIP+8(%rsp)
 	je	.Lbstep_iret
-
-	HANDLE_INTR_SAFERET 8(%rsp)
-
 	cmpq	$.Lgs_change, RIP+8(%rsp)
 	jne	.Lerror_entry_done_lfence
 
diff --git a/arch/x86/include/asm/nospec-branch.h b/arch/x86/include/asm/nospec-branch.h
index 150ac4ab5a02c..c49f42c8e5e18 100644
--- a/arch/x86/include/asm/nospec-branch.h
+++ b/arch/x86/include/asm/nospec-branch.h
@@ -87,50 +87,6 @@
 	add	$(BITS_PER_LONG/8), %_ASM_SP;		\
 	lfence;
 
-/*
- * Helper for detecting if an interrupt occurred at an unsafe location within
- * Safe-RET.  If Safe-RET is interrupted after the CALL or LEA the RSB may get
- * poisoned by the interrupt handler.
- *
- * The Safe-RET sequence is:
- *
- * CALL
- * LEA 8(%RSP), %RSP
- * RET
- *
- * The two CMPs below check whether RIP points to after the CALL or after the
- * LEA.
- *
- * The LFENCE below is to address this particular speculation case:
- *
- * 1. Userspace runs and poisons the BTB around the safe-RET routine
- *
- * 2. Userspace triggers some kind of exception
- *
- * 3. Kernel executes error_entry() and mis-speculates the branch into thinking
- *    it actually came from kernel space
- *
- * 4. The kernel then further mis-speculates that the exception occurred due
- *    to an interrupted safe-RET
- *
- * 5. The handle_interrupted_saferet() routine speculatively executes and
- *    speculatively does a safe-RET. But this is unsafe since it was never
- *    untrained.
- *
- * The LFENCE fixes this by ensuring step 5 is never reached speculatively.
- * Note that this LFENCE only occurs if safe-RET was actually interrupted (so
- * it's outside of the normal path).
- */
-#define __HANDLE_INTR_SAFERET(name, pt_regs)		\
-	cmpq	$(name), RIP+pt_regs;			\
-	jb	1f;					\
-	cmpq	$(name)+5, RIP+pt_regs;			\
-	ja	1f;					\
-	lfence;						\
-	leaq	pt_regs, %rdi;				\
-	call	handle_interrupted_saferet;		\
-	1:
-
 #ifdef __ASSEMBLY__
 
 /*
@@ -234,14 +190,6 @@
 #endif
 .endm
 
-.macro HANDLE_INTR_SAFERET pt_regs
-#ifdef CONFIG_MITIGATION_SRSO
-	ALTERNATIVE_2 "", \
-	__stringify(__HANDLE_INTR_SAFERET(srso_safe_ret, \pt_regs)), X86_FEATURE_SRSO, \
-	__stringify(__HANDLE_INTR_SAFERET(srso_alias_safe_ret, \pt_regs)), X86_FEATURE_SRSO_ALIAS
-#endif
-.endm
-
 /*
  * Macro to execute VERW insns that mitigate transient data sampling
  * attacks such as MDS or TSA. On affected systems a microcode update
@@ -501,10 +449,6 @@ static __always_inline void x86_idle_clear_cpu_buffers(void)
 		x86_clear_cpu_buffers();
 }
 
-void srso_safe_ret(void);
-void srso_alias_safe_ret(void);
-void handle_interrupted_saferet(struct pt_regs *regs);
-
 #endif /* __ASSEMBLY__ */
 
 #endif /* _ASM_X86_NOSPEC_BRANCH_H_ */
diff --git a/arch/x86/kernel/cpu/bugs.c b/arch/x86/kernel/cpu/bugs.c
index cdd85889968b1..57ba697e29180 100644
--- a/arch/x86/kernel/cpu/bugs.c
+++ b/arch/x86/kernel/cpu/bugs.c
@@ -3238,42 +3238,3 @@ ssize_t cpu_show_vmscape(struct device *dev, struct device_attribute *attr, char
 	return cpu_show_common(dev, attr, buf, X86_BUG_VMSCAPE);
 }
 #endif
-
-#ifdef CONFIG_MITIGATION_SRSO
-/*
- * Called during exception/interrupt entry if interrupted during the
- * safe-RET sequence.  The safe-RET sequence consists of 3 instructions:
- *
- *	CALL
- *	LEA 8(%RSP), %RSP
- *	RET
- *
- * An interrupt after the CALL or after the LEA could potentially lead
- * to branch predictor poisoning and results in the sequence not being
- * able to be safely resumed.
- *
- * Therefore, modify the regs state as if the remaining part of the
- * safe-RET sequence executed so the interrupt returns back to the
- * desired return target, instead of the to the safe-RET sequence.
- */
-void noinstr handle_interrupted_saferet(struct pt_regs *regs)
-{
-	unsigned long rip = regs->ip;
-
-	if (rip == (unsigned long) srso_safe_ret ||
-	    rip == (unsigned long) srso_alias_safe_ret) {
-	    /* Modify stack pointer as if LEA executed: */
-	    regs->sp += 8;
-	}
-
-	/*
-	 * Adjust registers as if RET executed:
-	 *
-	 * 1. Read the return address off the stack and into rIP:
-	 */
-	regs->ip = *(unsigned long *)(regs->sp);
-
-	/* 2. Pop rIP off the stack: */
-	regs->sp += 8;
-}
-#endif /* CONFIG_MITIGATION_SRSO */
diff --git a/arch/x86/lib/retpoline.S b/arch/x86/lib/retpoline.S
index 6d64bc69a6ec3..01fcf0cd679bd 100644
--- a/arch/x86/lib/retpoline.S
+++ b/arch/x86/lib/retpoline.S
@@ -113,24 +113,10 @@ __EXPORT_THUNK(srso_alias_untrain_ret)
 #endif
 
 SYM_START(srso_alias_safe_ret, SYM_L_GLOBAL, SYM_A_NONE)
-
-	/*
-	 * Tell objtool that those are not function pointers referenced by
-	 * __HANDLE_INTR_SAFERET(). Below too.
-	 */
-	ANNOTATE_NOENDBR
-
-	/*
-	 * Safe-RET sequence. If you need to change it, adjust
-	 * handle_interrupted_saferet() too.
-	 */
 	lea 8(%_ASM_SP), %_ASM_SP
 	UNWIND_HINT_FUNC
-
-	ANNOTATE_NOENDBR
 	ANNOTATE_UNRET_SAFE
 	ret
-	/* End of Safe-RET sequence */
 	int3
 SYM_FUNC_END(srso_alias_safe_ret)
 
@@ -245,14 +231,8 @@ SYM_START(srso_untrain_ret, SYM_L_GLOBAL, SYM_A_NONE)
  * the stack.
  */
 SYM_INNER_LABEL(srso_safe_ret, SYM_L_GLOBAL)
-	/*
-	 * Safe-RET sequence. If you need to change it, adjust
-	 * handle_interrupted_saferet() too.
-	 */
 	lea 8(%_ASM_SP), %_ASM_SP
 	ret
-	/* End of Safe-RET sequence */
-
 	int3
 	int3
 	/* end of movabs */
-- 
cgit 1.3-korg

